sql/src/main/java/io/crate/action/sql/SQLOperations.java

```diff
@@ -298,114 +285,24 @@ public void sync(CompletionListener listener) {
             LOGGER.debug("method=sync");
 
             if (throwable == null) {
-                /**
-                 * Execution is delayed to sync() to be able to make efficient bulk requests.
-                 *
-                 * JDBC batch execution can cause the following flow:
-                 *
-                 * Parse -> Bind -> Execute
-                 * Parse -> Bind -> Execute
-                 * Sync
-                 *
-                 * If the second parse is called with a *different* statement, the parse call will trigger sync.
-                 * Cleanup needs to happen before parse returns because clients don't wait for parseOk but follow up
-                 * with bind/execute immediately.
-                 *
-                 * This is why the global state is assigned to local variables and the cleanup is called immediately.
-                 * Otherwise there would be concurrency/async state modification.
-                 */
                 Portal portal = getSafePortal(lastPortalName);
-                if (portal.resultReceivers.size() == 1) {
-                    Plan plan = planner.plan(portal.analysis, jobId, 0, portal.maxRows);
-                    ResultReceiver resultReceiver = portal.resultReceivers.get(0);
-                    if (portal.maxRows > 0) {
-                        portal.resultReceivers.clear();
-                    }
-                    cleanup(lastPortalName);
-                    applySessionSettings(plan);
-                    resultReceiver.addListener(listener);
-                    resultReceiver.addListener(new StatsTablesUpdateListener(jobId, statsTables));
-
-                    if (!portal.analysis.analyzedStatement().isWriteOperation()) {
-                        resultReceiver = new ResultReceiverRetryWrapper(resultReceiver, this, portal, analyzer,
-                            planner, transportKillJobsNodeAction);
-                    }
-
-                    if (resumeIfSuspended(portal, resultReceiver)) return;
-                    RowReceiverToResultReceiver rowReceiver = new RowReceiverToResultReceiver(resultReceiver, portal.maxRows);
-                    portal.rowReceiver = rowReceiver;
-                    executor.execute(plan, rowReceiver);
-                } else {
-                    Object[][] bulkArgs = toBulkArgs(portal.bulkParams);
-                    Analysis analysis = analyzer.analyze(statement, new ParameterContext(new Object[0], bulkArgs, defaultSchema));
-                    ImmutableList<ResultReceiver> resultReceivers = ImmutableList.copyOf(portal.resultReceivers);
-                    Plan plan = planner.plan(analysis, jobId, 0, portal.maxRows);
-                    cleanup(lastPortalName);
-                    executeBulk(plan, resultReceivers, listener);
-                }
+                Plan plan = portal.prepareSync(planner);
+                applySessionSettings(plan);
+                portal.sync(analyzer, planner, statsTables, listener);
```

Comment:
portal already has a reference to the analyzer so I think the argument here could be removed?